Introduce deprecation policy to developer guide. (#10252)
rapidsai/cudf: cpp/docs/DEVELOPER_GUIDE.md

[**Anonymous namespaces should *never* be used in a header file.**](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file)

# Deprecating and Removing Code

> If changes to libcudf involve breaking existing APIs, libcudf follows the following policy:

Reviewer: Maybe there's more to say about this. One kind of breaking change is when we remove an API, and this text explains well what should be done in that case. More often, we replace an API, and it's not clear from this text at which point the new API should be added.
Also might be good to mention that the new API should be referenced in the deprecation message.

Author: I've rewritten this to add some more information. Let me know what you think.

- Code should be deprecated for one release prior to removal.
- Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.

> - Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.

Author: This technically should go in DOCUMENTATION.md, not here, but I figured it was a single line and would be more discoverable here. I can add something to DOCUMENTATION.md in addition (or instead) and link here if reviewers think there would be a benefit.

Reviewer 2: I think it should go in both. Please also add an example of using the doxygen `deprecated` command to DOCUMENTATION.md

Author: Done. I also added it to an API locally and built the docs to verify that it worked as expected. Conveniently, in addition to showing up on the documentation for a particular API it looks like doxygen will generate a page containing all deprecations:

![image](https://user-images.githubusercontent.com/1538165/154392371-cea3166d-8f6d-4314-a425-a4e7116f6362.png)

![image](https://user-images.githubusercontent.com/1538165/154392324-4a201c76-d626-4e5f-8cee-c04ac2ad4cd5.png)

PRs containing deprecations should be labeled as "non-breaking".

> The subsequent pull request in which deprecated code is moved should be labeled as "breaking".

Author: I don't know if we need to document this and can remove if we think it's extraneous. I figured it would be easier to have it here to start in case it merits discussion.

Reviewer 3: I think the "breaking" policy is valuable to include. It is helpful to know and would be unclear without this note.

[**Anonymous namespaces should *never* be used in a header file.**](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file)

# Deprecating and Removing Code

To minimize disruption for users, any breaking changes must be signaled by deprecation warnings for one release prior to the change.
This includes wholesale removals of functions/methods/classes *and* changes to existing function signatures.
Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.
Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.

When an API is deprecated, if a replacement exists the replacement should be mentioned in both the deprecation message and the deprecation documentation.
Changes to function parameters should be accompanied by the introduction of suitable overloads to avoid API breaks for one release cycle.

> Where possible, deprecated APIs should call new APIs, but duplication in deprecated code is acceptable at the discretion of PR reviewers if avoiding it imposes an undue burden on developers.

Reviewer 2: Please adjust your word wrapping -- suggest keeping it to the same as our .cpp files -- 100 characters.

Author: Done.

> PRs containing deprecations should be labeled as "non-breaking".

Reviewer 4: This is where the whole "breaking" vs. "non-breaking" thing starts to "break down" :) 

Using a `[[deprecated]]` keyword on an API will cause a compiler warning by default. If a user has warnings as errors enabled, their build is broken until they disable that warning or update to the new API. Was that a breaking change?

Author: I chose to ignore this issue in what I wrote up. Feel free to suggest an alternative if you have a suggestion on what might be better to write here. I think using `[[deprecated]]` is still better than not, since at least for people not compiling with `Wall` we'll be offering a gentler off ramp. For people who are, the situation is no different than an outright removal.

# Deprecating and Removing Code

libcudf is constantly growing and improving to improve performance and better meet our users' needs.
As a result, we occasionally need to break or entirely remove APIs to respond to new and improved
understanding of the functionality we provide. Remaining free to do this is essential to making
libcudf an agile library that can rapidly accommodate our users needs. We encourage users to
[Live At Head](https://www.youtube.com/watch?v=tISy7EJQPzI). On a best effort basis, the libcudf
team will notify users of changes that we expect to have significant or widespread effects.

Where possible, pending API removals will be indicated using the
[deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute and
documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated)
command prior to removal. When an API is deprecated, if a replacement exists the replacement should
be mentioned in both the deprecation message and the deprecation documentation. The PRs introducing
deprecations should also be labeled "deprecation" to facilitate discovery and removal in the
subsequent release.

> In order to advertise breaking changes, any PR that breaks or removes an existing API should be
> tagged as "breaking". This will ensure that the "Breaking" section of the release notes will include
> a description of what has broken from the past release. Note that PRs containing deprecations should
> be labeled as "non-breaking".

Reviewer 2: ```suggestion
Advertise breaking changes by labeling any pull request that breaks or removes an existing API with the "breaking" tag. This ensures that the "Breaking" section of the release notes includes
a description of what has broken from the past release. Label pull requests that contain deprecations with the "non-breaking" tag.
```

# Deprecating and Removing Code

libcudf is constantly evolving to improve performance and better meet our users' needs.
As a result, we occasionally need to break or entirely remove APIs to respond to new and improved
understanding of the functionality we provide. Remaining free to do this is essential to making
libcudf an agile library that can rapidly accommodate our users needs. We encourage users to
[Live At Head](https://www.youtube.com/watch?v=tISy7EJQPzI). On a best effort basis, the libcudf
team will notify users of changes that we expect to have significant or widespread effects.

Where possible, indicate pending API removals using the
[deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute and

> document them using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated)

Reviewer 2: Line length. I know this is annoying and manual, but the rest of the doc is wrapped at the same column. Wish clang-tidy would do markdown files. This doesn't mention anything about line length: https://github.com/notslang/tidy-markdown

Author: I think you looked at the specific commit where I pushed your changes and missed the commit afterwards where reformatted all line lengths. Can you take a look at the current version and make sure the line lengths look like you'd expect?
